File: app/src/main/java/com/patryk/springer/cashmanager/view/listdetails/ListDetailsFragment.java
package com.patryk.springer.cashmanager.view.listdetails;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.patryk.springer.cashmanager.R;
import com.patryk.springer.cashmanager.view.base.BaseFragment;
import com.patryk.springer.cashmanager.view.main.MainActivity;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

/**
 * Created by Patryk Springer on 2019-06-17.
 */
public class ListDetailsFragment extends BaseFragment<ListDetailsContract.Presenter>
        implements ListDetailsContract.View, ActionMode.Callback {

    static final int NEW_PRODUCT_INTENT_CODE = 8162;
    static final int EDIT_PRODUCT_INTENT_CODE = 9351;
    static final String PRODUCT_NAME_KEY = "productName";
    static final String PRODUCT_QUANTITY_KEY = "quantity";
    private static final String LIST_ID_KEY = "listId";
    @Inject
    ListDetailsContract.Presenter mPresenter;
    private FloatingActionButton mFabAddProduct;
    private ProductsListAdapter mAdapter;
    private ActionMode mActionMode = null;

    public static ListDetailsFragment newInstance(int listId) {
        ListDetailsFragment fragment = new ListDetailsFragment();
        Bundle bundle = new Bundle();
        bundle.putInt(LIST_ID_KEY, listId);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    protected int getLayoutResId() {
        return R.layout.fragment_shopping_details;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        RecyclerView rvShoppingList = view.findViewById(R.id.rv_shopping_details_list);
        mFabAddProduct = view.findViewById(R.id.fab_shopping_details_add_product);
        mAdapter = new ProductsListAdapter(mPresenter);
        rvShoppingList.setAdapter(mAdapter);
        rvShoppingList.setLayoutManager(new LinearLayoutManager(getContext()));

        mFabAddProduct.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showNewProductDialog();
            }
        });

        Bundle bundle = getArguments();
        if (bundle != null) {
            int listId = bundle.getInt(LIST_ID_KEY);
            mPresenter.refreshView(listId);
        }
    }

    private void showNewProductDialog() {
        NewProductDialogFragment dialog = new NewProductDialogFragment();
        dialog.setTargetFragment(this, NEW_PRODUCT_INTENT_CODE);
        dialog.show(getFragmentManager(), null);
    }


    @Override
    public void showListName(@NotNull String name, boolean isArchived) {
        String listName = name;
        if (isArchived) {
            listName += " [" + getString(R.string.archived_list) + "]";
        }
        ((MainActivity) getMBaseActivity()).setToolbarTitle(listName);
    }

    @Override
    public void showAddNewProductButton(boolean isVisible) {
        if (isVisible) {
            mFabAddProduct.show();
        } else {
            mFabAddProduct.hide();
        }
    }

    @Override
    public void refreshList() {
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void showActionMenu() {
        mActionMode = ((MainActivity) getMBaseActivity()).showActionMenu(this);
    }

    @Override
    public void showEditProductDialog(@NotNull String name, int quantity) {
        NewProductDialogFragment dialog =
                NewProductDialogFragment.Companion.newInstance(name, quantity);
        dialog.setTargetFragment(this, EDIT_PRODUCT_INTENT_CODE);
        dialog.show(getFragmentManager(), null);
    }

    @NotNull
    @Override
    public ListDetailsContract.Presenter getMPresenter() {
        return mPresenter;
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        MenuInflater inflater = mode.getMenuInflater();
        inflater.inflate(R.menu.product_action_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        int itemId = item.getItemId();
        if (itemId == R.id.menu_delete) {
            mPresenter.onProductRemoved();
            mode.finish();
            return true;
        } else if (itemId == R.id.menu_edit) {
            mPresenter.onProductEditClicked();
            mode.finish();
            return true;
        }
        return false;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        mActionMode = null;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            String name = data.getStringExtra(PRODUCT_NAME_KEY);
            int quantity = data.getIntExtra(PRODUCT_QUANTITY_KEY, -1);
            handleProductIntent(name, quantity, requestCode);
        }
    }

    private void handleProductIntent(String name, int quantity, int requestCode) {
        if (requestCode == EDIT_PRODUCT_INTENT_CODE) {
            mPresenter.onProductEditConfirmed(name, quantity);
        } else if (requestCode == NEW_PRODUCT_INTENT_CODE) {
            mPresenter.onNewProductAdded(name, quantity);
        }
    }
}

File: app/src/main/java/com/patryk/springer/shoppinglist/view/listdetails/ListDetailsFragment.kt
package com.patryk.springer.shoppinglist.view.listdetails

import android.content.DialogInterface
import android.os.Bundle
import android.view.ActionMode
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import com.patryk.springer.shoppinglist.R
import com.patryk.springer.shoppinglist.view.base.BaseFragment
import com.patryk.springer.shoppinglist.view.main.MainActivity
import kotlinx.android.synthetic.main.dialog_create_product.view.*
import kotlinx.android.synthetic.main.fragment_shopping_details.*
import javax.inject.Inject

/**
 * Created by Patryk Springer on 2019-06-15.
 */

class ListDetailsFragment : BaseFragment<ListDetailsContract.Presenter>(), ListDetailsContract.View,
		ActionMode.Callback {


	override val layoutResId: Int
		get() = R.layout.fragment_shopping_details
	@Inject
	override lateinit var mPresenter: ListDetailsContract.Presenter
	private val mAdapter: ProductsListAdapter by lazy { ProductsListAdapter(mPresenter) }
	private var mActionMode: ActionMode? = null

	companion object {
		private const val LIST_ID_KEY: String = "listId"

		fun newInstance(listId: Int): ListDetailsFragment {
			val fragment: ListDetailsFragment = ListDetailsFragment()
			val bundle: Bundle = Bundle()
			bundle.putInt(LIST_ID_KEY, listId)
			fragment.arguments = bundle
			return fragment
		}
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)
		rv_shopping_details_list.adapter = mAdapter
		rv_shopping_details_list.layoutManager = LinearLayoutManager(context)

		fab_shopping_details_add_product.setOnClickListener {
			showNewProductDialog()
		}

		arguments?.let { bundle ->
			val listId: Int = bundle.get(LIST_ID_KEY) as Int
			mPresenter.refreshView(listId)
		}
	}

	private fun showNewProductDialog() {
		val view = LayoutInflater.from(context).inflate(R.layout.dialog_create_product, null)
		val positiveCallback: DialogInterface.OnClickListener =
			DialogInterface.OnClickListener { dialog, which ->
				val etNewProductName =
					(dialog as AlertDialog).findViewById<EditText>(R.id.et_create_product_name)
				val name = etNewProductName?.text.toString()
				val etNewProductQuantity =
					dialog.findViewById<EditText>(R.id.et_create_product_quantity)
				val quantity: Int = etNewProductQuantity?.text?.toString()?.toIntOrNull() ?: 0
				if (name.isNotBlank() && quantity > 0) {
					mPresenter.onNewProductAdded(name, quantity)
				}
			}
		showAlertDialog(R.string.create_new_product_title, view, positiveCallback)
	}

	override fun showAddNewProductButton(isVisible: Boolean) {
		if (isVisible) {
			fab_shopping_details_add_product.show()
		} else {
			fab_shopping_details_add_product.hide()
		}
	}

	override fun showListName(name: String, isArchived: Boolean) {
		val listName = if (isArchived) {
			"$name [${getString(R.string.archived_list)}]"
		} else name
		(mBaseActivity as? MainActivity)?.setToolbarTitle(listName)
	}

	override fun refreshList() {
		mAdapter.notifyDataSetChanged()
	}

	override fun showActionMenu() {
		mActionMode = (mBaseActivity as? MainActivity)?.showActionMenu(this)
	}

	override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?): Boolean {
		return when (item?.itemId) {
			R.id.menu_delete -> {
				mPresenter.onProductRemoved()
				mode?.finish()
				true
			}
			R.id.menu_edit -> {
				mPresenter.onProductEditClicked()
				mode?.finish()
				true
			}
			else -> false
		}
	}

	override fun showEditProductDialog(name: String, quantity: Int) {
		val view = LayoutInflater.from(context).inflate(R.layout.dialog_create_product, null)
		view.et_create_product_name.setText(name)
		view.et_create_product_quantity.setText(quantity.toString())
		val positiveCallback: DialogInterface.OnClickListener =
			DialogInterface.OnClickListener { dialog, which ->
				val etNewProductName =
					(dialog as AlertDialog).findViewById<EditText>(R.id.et_create_product_name)
				val name = etNewProductName?.text.toString()
				val etNewProductQuantity =
					dialog.findViewById<EditText>(R.id.et_create_product_quantity)
				val quantity: Int = etNewProductQuantity?.text?.toString()?.toIntOrNull() ?: 0
				if (name.isNotBlank() && quantity > 0) {
					mPresenter.onProductEditConfirmed(name, quantity)
				}
			}
		showAlertDialog(R.string.edit_product_title, view, positiveCallback)
	}

	override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
		val inflater = mode?.menuInflater
		inflater?.inflate(R.menu.product_action_menu, menu)
		return true
	}

	override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean {
		return false
	}

	override fun onDestroyActionMode(mode: ActionMode?) {
		mActionMode = null
	}
}